pkg/reconciler/buildrun/resources/service_accounts.go

```diff
@@ -38,70 +39,114 @@ func IsGeneratedServiceAccountUsed(buildRun *buildv1alpha1.BuildRun) bool {
 	return buildRun.Spec.ServiceAccount != nil && buildRun.Spec.ServiceAccount.Generate
 }
 
+// GenerateSA generates a new service account on the fly
+func GenerateSA(ctx context.Context, client client.Client, build *buildv1alpha1.Build, buildRun *buildv1alpha1.BuildRun) (serviceAccount *corev1.ServiceAccount, err error) {
+	serviceAccount = &corev1.ServiceAccount{}
+	err = client.Get(
+		ctx,
+		types.NamespacedName{
+			Name:      GetGeneratedServiceAccountName(buildRun),
+			Namespace: buildRun.Namespace},
+		serviceAccount)
+
+	switch {
+	case err == nil: // if the service account already exists, do nothing and just return it
+		ctxlog.Info(ctx, "serviceAccount for BuildRun already exists", namespace, buildRun.Namespace, name, serviceAccount.Name, "BuildRun", buildRun.Name)
+		return serviceAccount, nil
+
+	case apierrors.IsNotFound(err):
+		// populate the object with well known required fields
+		serviceAccount = &corev1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      GetGeneratedServiceAccountName(buildRun),
+				Namespace: buildRun.Namespace,
+				Labels:    map[string]string{buildv1alpha1.LabelBuildRun: buildRun.Name},
+				OwnerReferences: []metav1.OwnerReference{
+					*metav1.NewControllerRef(buildRun, buildv1alpha1.SchemeGroupVersion.WithKind("BuildRun")),
+				},
+			},
+			AutomountServiceAccountToken: pointer.BoolPtr(false),
+		}
+		ctxlog.Debug(ctx, "automatic generation of service account", namespace, serviceAccount.Namespace, name, serviceAccount.Name)
+
+		// add the secrets references into the new sa
+		ApplyCredentials(ctx, build, serviceAccount)
+
+		// if we didnt have the sa, then create one and ensure the Build secrets are referenced
+		if err := client.Create(ctx, serviceAccount); err != nil {
+			return nil, err
+		}
+
+		ctxlog.Info(ctx, "created serviceAccount for BuildRun", namespace, buildRun.Namespace, name, serviceAccount.Name, "BuildRun", buildRun.Name)
+		return serviceAccount, nil
+
+	default:
+		return nil, err
+	}
+}
+
+// getDefaultNamespaceSA retrieves a pipeline or default sa per namespace. This is used when users do not specify a service account
+// to use on BuildRuns
+func getDefaultNamespaceSA(ctx context.Context, client client.Client, buildRun *buildv1alpha1.BuildRun) (*corev1.ServiceAccount, error) {
+	serviceAccount := &corev1.ServiceAccount{}
+
+	err := client.Get(ctx, types.NamespacedName{Name: PipelineServiceAccountName, Namespace: buildRun.Namespace}, serviceAccount)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	} else if apierrors.IsNotFound(err) {
+		ctxlog.Info(ctx, "falling back to default serviceAccount", namespace, buildRun.Namespace)
+		err = client.Get(ctx, types.NamespacedName{Name: DefaultServiceAccountName, Namespace: buildRun.Namespace}, serviceAccount)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return serviceAccount, nil
```

Comment:
I did it here on the spot, but I think it can be simplified to something like this:
```suggestion
	if err != nil && apierrors.IsNotFound(err) {
		ctxlog.Info(ctx, "falling back to default serviceAccount", namespace, buildRun.Namespace)
		err = client.Get(ctx, types.NamespacedName{Name: DefaultServiceAccountName, Namespace: buildRun.Namespace}, serviceAccount)
	}

	return serviceAccount, err
```

Reply:
Thanks, I like the simplicity but also I prefer to make it very obvious on whats happening. Anyways this code will be removed after this PR, because of https://github.com/shipwright-io/build/issues/680 . If u dont mind I would like to do via another PR.